### PR TITLE
Backward history powered snapshots

### PR DESCRIPTION
**Pull request by n1k0**
r=? @leplatrem

### CODE REVIEW

File: src/collection.js [hunk 1]
```diff
@@ -337,32 +375,23 @@ export default class Collection {
...
-            snapshot = snapshot.filter(r => r.id !== record.id);
-          } else if (!seenIds.has(record.id)) {
-            seenIds.add(record.id);
-            snapshot.push(record);
-          }
-        }
+    });
+    return Promise.all([listRecords, listHistory])
+      .then(([{data: records}, {data: changes}]) => {
+        const snapshot = computeSnapshotAt(at, records, changes);
```
Comment (leplatrem) [inline src/collection.js]: 👍

File: test/collection_test.js [hunk 2]
```diff
@@ -420,3 +420,15 @@ describe("Collection", () => {
     });
   });
 });
+
+describe("computeSnapshotAt()", () => {
+  // Note: most of the tests covering this API are done in integration_test.js
+
+  it("should raise when a snapshot records timestamp exceeds requirements", () => {
```
Comment (leplatrem) [inline test/collection_test.js]: nit: *exceeds period covered by changes* maybe?

File: test/collection_test.js [hunk 3]
```diff
@@ -420,3 +420,15 @@ describe("Collection", () => {
...
+    const records = [{id: 1, last_modified: 43}, {id: 2, last_modified: 44}];
+    const changes = [{action: "create", target: {data: {last_modified: 40}}}];
```
Comment (leplatrem) [inline test/collection_test.js]: If we want to simulate that the `history` plugin was enabled after the record `id: 1` was created, then we should have `last_modified: 44` (or 45, but above 43)

Imagine the following scenario:

- `id: 1` is created (ts: 43)
- plugin is enabled 
- `id: 2`is created (ts: 44), change entry `action: create` is tracked (ts >= 44)

File: test/collection_test.js [hunk 4]
```diff
@@ -420,3 +420,15 @@ describe("Collection", () => {
...
+
+    expect(() => computeSnapshotAt(42, records, changes))
```
Comment (leplatrem) [inline test/collection_test.js]: should be below oldest available change

File: src/collection.js [hunk 5]
```diff
@@ -6,6 +6,44 @@ import endpoint from "./endpoint";
 
 
 /**
+ * Computes a snapshot of a collection at a given timestamp, from a current list
+ * of records and a list of changes.
+ */
+export function computeSnapshotAt(at, records, changes) {
```
Comment (leplatrem) [inline src/collection.js]: meganit: I would have put `at` as last argument :)

**Comment by n1k0** [comment]
After more checking, it looks like history deletion entries only have a tombstone as a target:



That makes this patch unable to deal with restoring deleted entries I think. Thoughts?

**Comment by n1k0** [comment]
Updated implementation in https://github.com/Kinto/kinto-http.js/pull/163/commits/71924f60abc90307765060ad487be408e4b99ebe. I think we should carefully review the tests, possibly add more to ensure we're not missing any edge case.

**Comment by n1k0** [comment]
Ok, during my latest investigation I've stumbled upon a limitation in ongoing implementation where updated records weren't properly picked by the backward algorythm. I've pushed new tests on master in https://github.com/Kinto/kinto-http.js/commit/585d5518f670469c112944ffcc5c2e21fe356f3d to ensure it was working with the previous one, so they're failing now with this branch.

We'll have to fix this. TOMORROW. 'night.

**Comment by n1k0** [comment]
Funnily enough we don't have unit and integration tests running anymore on Travis:

**Comment by n1k0** [comment]
Tests enabled again. Kinto server setup now needs proper fixing

**Comment by n1k0** [comment]
Ok at least we now have our test suite running again, and we have the two errors we should fix with our implementation (plus another one not related to this patch):

**Comment by n1k0** [comment]
Note that I've created a separate PR to track the other failing test issue in https://github.com/Kinto/kinto-http.js/pull/164

**Comment by leplatrem** [comment]
Can't be achieved with the current history API since history entries contain the current (ie. new) version of the object during the operation.
